aptos-move/framework/aptos-framework/sources/Token.move

```diff
@@ -476,175 +490,288 @@ module AptosFramework::Token {
+            0, // haven't minted token yet
+            maximum,
+            uri,
+            royalty_payee_address,
+            royalty_points_denominator,
+            royalty_points_nominator,
+            token_mut_config,
+            property_keys,
+            property_values,
+            property_types
         );
-        Table::add(
-            &mut collections.mint_capabilities,
-            token_id,
-            MintCapability { token_id },
+
+        let token_id = mint_token(
+            account,
+            tokendata_id,
+            balance,
+            Vector::empty<String>(),
+            Vector::empty<vector<u8>>(),
+            Vector::empty<String>()
         );
 
-        if (initial_balance > 0) {
-            initialize_token_store(account);
-            initialize_token(account, token_id);
-            mint(account, Signer::address_of(account), token_id, initial_balance);
+        token_id
+    }
+
+    fun get_properties_to_be_updated(
+        token_data: &TokenData,
+        keys: &vector<String>,
+        values: &vector<vector<u8>>,
+        types: &vector<String>
+    ): PropertyList {
+        let res_keys = Vector::empty<String>();
+        let res_vals = Vector::empty<vector<u8>>();
+        let res_types = Vector::empty<String>();
+        let default_properties = &token_data.default_property_list;
+        let i = 0;
+        while (i < Vector::length(keys)) {
+            let k = Vector::borrow(keys, i);
+            let v = Vector::borrow(values, i);
+            let t = Vector::borrow(types, i);
+            if (PropertyList::contains(default_properties, k) ) {
+                if ( PropertyList::get_type(default_properties, k) == *t && PropertyList::get_value(default_properties, k) != *v ) {
+                    Vector::push_back(&mut res_keys, *k);
+                    Vector::push_back(&mut res_vals, *v);
+                    Vector::push_back(&mut res_types, *t);
+                };
+            };
+            i = i + 1;
         };
+        PropertyList::new(&res_keys, &res_vals, &res_types)
+    }
+
+    public fun mint_token(
+        account: &signer,
+        token_data_id: TokenDataId,
+        amount: u64,
+        keys: vector<String>,
+        values: vector<vector<u8>>,
+        types: vector<String>
+    ): TokenId acquires Collections, TokenStore {
+        /*
+        TODO: capability and signer check
+        */
+        assert!(Vector::length(&keys) == Vector::length(&values), 1);
+        assert!(Vector::length(&types) == Vector::length(&values), 1);
+
+        let creator_addr = token_data_id.creator;
+        let all_token_data = &mut borrow_global_mut<Collections>(creator_addr).token_data;
+        let token_data = Table::borrow_mut(all_token_data, token_data_id);
+
+        assert!(token_data.supply + amount <= token_data.maximum, 1);
+
+        token_data.supply = token_data.supply + amount;
+        token_data.largest_serial_number = token_data.largest_serial_number + 1;
```

Comment:
I think one of the goals was to support either serial numbers or supply but not both as it makes the logic too complex and the applications seem to be binary between the two.

Reply:
I delete the serial number for now. 
For application, how do you handle the case if you create 1000 token and you mutate the property value of 500 tokens to a different value? 

Based on original purpose, I think serial_number is just for grouping token based on the property value.
if we can guarantee if tokens of the same serial number have the same property_value, that is good enough.  I don't want to add constraint that they should create a new serial_number for each token. 

I also feel the current implementation is a simple extension of the previous data model (only add serial_number field).